README.md
# Ugl-server

The server side code of project Ugl.

# Table of Contents

 - [Introduction](#introduction)
  - [Architecture](#architecture)
  - [Files](#files)
  - [Keys](#keys)
  - [Planning](#planning)
 - [Server API Notes](#server-api-notes)
  - [Types of Responses](#types-of-responses)
     - [Success](#1-success)
     - [Error](#2-error)
  - [Encryption](#encryption)
     - [One-way Encryption](#1-one-way-encryption)
     - [Two-way Encryption](#2-two-way-encryption)
 - [API Events](#api)
	 - [User Identities](#1-user-identities)
		 - [login](#1-login)
		 - [logout](#2-logout)
		 - [register](#3-register)
		 - [revokeToken](#4-revoketoken)
		 - [oauth_clientCallback](#5-oauth_clientcallback)
		 - [Forgot password](#6-forgot_password)
		 - [Change password](#9-change_password)
		 - [Get profile of a user](#7-get-profile-of-a-user)
		 - [Set profile for a user](#8-set-profile-for-a-user)
	 - [Group API](#2-group-api)
		 - [Get the groups a user joins](#1-list-groups-of)
		 - [Get the profile of a aroup](#6-get-the-profile-of-a-group)
		 - [Create a group](#2-create-a-group)
		 - [Delete, leave a group, or kick members](#3-delete-or-leave-a-group-or-kick-members)
		 - [Transfer ownership](#4-transfer-ownership-of-a-group)
		 - [Edit group profile](#5-edit-group-profile)
	 - [News](#3-news)

***

# Introduction

## Architecture
Basic features:

* written in PHP 5.5+ and run on Nginx 1.5+
* follows MVC model with autoloader
* generic caching mechanism
* generic database class

Notes:

* keep an eye on facebook's HipHop VM
* Xiangyu owns the root of the server so the infrastructure can be changed when needed
* Web Client front-end is written in HTML5 and CSS3 built ono top of Bootstrap framework.

## Files
 * `app` (0750) stores the libraries, configuration files, controllers and models
	 * `controllers` stores the controllers
	 * `models` saves all the model files
	 * `views` stores the views (templates)
 * `assets` (0750) has the public accessible files like images, css, and javascript
 * `data` (0774) stores logs and files uploaded by users
	 * `log` (0755) system log
	 * `upload` (0774) user upload files
 * `tmp` if exists, will store temporary files like cache
 * `vendor` if exists, stores reference libraries

## Keys
 * Mail Server:
     * Refer to http://windows.microsoft.com/en-us/windows/outlook/send-receive-from-app
	 * Username: ugl@sige.us
	 * Password: Boilermaker!

***

# Server API Notes

## Types of responses

All responses, regardless of types, will have header attribute of HTTP status **200**.

There are two types of responses:

### 1. Success

Server will respond to the client with json text of structure like:
```javascript
{
    "status": "success",
    "expiration": "2014-02-16T20:58:30+00:00",
    "data": {
    }
}
```
where

* `status` will always be `success` for successful responses.
* `expiration` is the time when this response becomes obsolete and must be re-fetched. It can be used for constructing client caching mechanism. The datetime format conforms to ISO-8601 date standard (see the example).
* The `data` property contains the actual data requested, which differs from request to request.

### 2. Error

```javascript
{
    "status": "error",
    "error": 1,
    "message": "Invalid email"
}
```
where

* `status` is always `error` for this type.
* `error` is the error number associated. One can refer to the document of a particular event for what the error number stands for.
* `message` is the message (pre-defined description) returned by the server.

The client should be able to distinguish between these two types of messages by the `status` field.

## Encryption

### 1. One-way Encryption

All non-decipherable passwords are expected to be encrypted in the following way before sending to the server:

* S1: Use SHA-1 to encode the string. Eg., `123456` becomes `7c4a8d09ca3762af61e59520943dc26494f8941b`
* S2: Use SHA-1 to encode the string got from S1. Eg., `69c5fcebaa65b560eaf06c3fbeb481ae44b8d618`
* S3: Base64 the string from S2. Eg., `NjljNWZjZWJhYTY1YjU2MGVhZjA2YzNmYmViNDgxYWU0NGI4ZDYxOA==`
* S4: To save space, use MD5 to encode the string from S3. Eg., `aaf53a928ca9baa6df03a5fe6e3c7b71`

In short, `$str = md5(base64_encode(sha1(sha1($str))));`

Client can make this a function so that it can be called conveniently.

### 2. Two-way encryption

All data that contains critical information and needs to be decoded should be encrypted in the following way before sending out:

* S1: Organize the information in the specified way to a string (event-dependent)
* S2: Use **AES-256-ECB** to encrypt the string got from step 1
	 * For android client, the private key to encrypt string should be `2IwehG2VEm3WhjLRMK/1aUPqAdW7KNvvRuskedxuOgOQ2jbO+wkKs5p5qJwh98GM`.
	 * Reference: http://stackoverflow.com/questions/10451068/encryption-mismatch-between-java-and-php
* S3: Encode the data from step 2 with Base64 and URLencode so that it can be sent out via HTTP protocol.
* S4: **If the method to send data is POST and this encryption is used, android client should post field `from=ugl_android`; web client should post field `from=ugl_web`.

In short `$str=urlencode(base64_encode(aes256($str, $key)))`

***

# API

## 1. User Identities

The data fields for User model are defined as below:

* __id__ (int): the unique identifier for a user.
* __email__ (string): the email of the user, conforming to RFC 2822 standard but with a maximum length of 200 chars. Required.
* ___password__ (string): the password hash (not available to clients).
* __nickname__ (string): HTML-filtered-out string with maximum length of 100 chars.
* __first_name__ (string): HTML-filtered-out string with maximum length of 100 chars. Required.
* __last_name__ (string): HTML-filtered-out string with maximum length of 100 chars. Required.
* __avatar_url__ (string): The URL of the avatar image with max length of 300 chars.
* __phone__ (string): Phone number. At most 36 chars.
* __description__ (string): HTML-encoded string. At most 150 chars.
* ___preferences__ (array): The preferences of the user
	 * `autoAcceptInvitation`
		 * `true`: accept group invitations automatically
		 * `false` (default): send me an email when I receive invitations
	 * `showMyProfile`
		 * `true` (default): allow others to see my profile
		 * `false`: do not allow
	 * `showMyPublicGroups`
		 * `true` (default): allow others to view the **public** groups I am in
		 * `false`: no one can see what groups I am in

Notes:

* HTML-filtered-out string means removing `<`, `>`, `;`, `"`, `\n`, `\t` from the string.

### 1) login
Log a user in. **To be updated to reflect token-based system.**

#### Request

| Name   | Description                        |
| ------ | ---------------------------------- |
| Method | POST                               |
| URL    | `api/login`                        |
| DATA   | `email`=abc@abc.cc&`password`=pass&`from`=ugl_android |

Note:

* For other clients, `from` should be changed to the corresponding client identifier.

**Sanity check:**
* `email` is a valid email address (RFC 2822)
* `password` equals `pass` which is the one-way encrypted string of the original 
   password. The original password must be at least 6 chars long (server cannot check).

#### Response
For example,
```javascript
{
    "status": "success",
    "expiration": "2014-02-16T20:58:30+00:00",
    "data": {
		"user_id": 12345,
		"ugl_token": something
    }
}
```

Notes:

* The client should save the `user_id` and `ugl_token` and send them back to the server 
when requesting information so that the server can identify the user as logged in. Think 
of the `ugl_token` as a special, volatile password associated with the user.
* After first logging in, client should fetch the profile of the user and cache the profile array.
	* When the profile array data gets changed, server will return a new profile array for the client to update local cache.

#### Associated Errors
* 100 - Email or password not provided
* 101 - Invalid email address
* 102 - User not found, or email and password do not match
* 103 - Password should not be empty
* 104 - Your account is temporarily on held for security concern. Please retry later or use social account to log in (too many failed attempts)

### 2) logout
Logout is a web client-only event. For mobile apps please use [revokeToken](#5-revoketoken) event.

#### Associated Errors

### 3) register
Register an account. **To be updated to reflect token-based system.**

#### Request
**Format:**

* Method: POST
* URL: `api/register`
* Data: `from`=ugl_android&`email`=something&`password`=pass&`confirm_pass`=pass&`first_name`=aaa&`last_name`=bbb&`agree`=true

Note:

* For other clients, `from` should be changed to the corresponding client identifier.

**Sanity check:**

* `email` is a valid email address (RFC 2822)
* `password` must be at least 6 chars long (server cannot check)
* `password` and `confirm_pass` must match
* `first_name` and `last_name` must not be empty
* `first_name` and `last_name` must not contain any one of the following chars: `<`, `>`, `;`, `"`, `\n`, `\t`, TBA.
* `agree` must be `true`

**Encryption:**

refer to the Encryption section.

#### Response
Same as the log in response.

#### Associated Errors
* 100 - Email, password, or name not provided
* 101 - Invalid email address
* 102 - Password and confirm password do not match
* 103 - First name or last name should be non-empty words (either empty or contain special chars)
* 104 - Email already registered
* 105 - You must agree to the terms of services to sign up (field "agree" != "true")
* 106 - Password should be at least 6 chars

### 4) revokeToken
Revoke the token used currently.

#### Request
**Format:**

* Method: POST
* URL: `api/revokeToken`
* Data: `user_id`=myid&`ugl_token`=mytoken

**Sanity check:**
none.

#### Response
A typical success message with data->message being "Token has been revoked.".

**Note:**
* This event is used to log out a user.
* Only when the user id and token really match will the token be reset.

#### Associated Errors
* 1 - Authentication fields missing (`user_id` or `ugl_token` is empty or null)
* 2 - User id should be a number (`user_id` is not a numeric value)

### 5) oauth_clientCallback
**Under Construction**

native client-only API used for telling server that a user successfully authenticates 
the app with the oauth provider.

#### Request

Organize the oauth response information to json text of the following format:

```javascript
{
    "auth": {
        "provider": "The provider with which the user authenticated, eg., \"Facebook\"",
        "uid": "An identifier unique to the given provider, such as a Twitter user ID.",
        "info": {
            "name": "The display name of the user given by the provider. Usually a nickname or firstname",
            "email": "The email of the authenticating user.",
            "nickname": "nickname of the user returned by the provider",
            "first_name": "",
            "last_name": "",
            "location": "",
            "description": "",
            "image": "",
            "phone": "",
            "urls": {
                "facebook": "Eg., http:\/\/facebook.com\/uzyn.chua returned by Facebook",
                "website": "Eg., http:\/\/xybu.me"
            }
        },
        "credentials": {
            "token": "Supplied by OAuth and OAuth 2.0 providers, the access token.",
            "secret": "Supplied by OAuth providers, the access token secret."
        }
    },
    "timestamp": "Time (in ISO 8601 format) when this auth was prepared.",
    "signature": "see below"
}
```

where

* `signature` should be the SHA-128 code of the string of the `auth` part above.
* `timestamp` has format of `2014-03-04T23:20:28+00:00` (ISO 8601)
* other fields should be self-evident given the information returned by the server.

And then use two-way encryption to encode the json text above. Let $data denote the encrypted data.

* Method: POST
* URL: /api/oauth/callback
* Data: `from`=ugl_android&`data`=$data

#### Response
Same as the log in response.

#### Associated Errors
TBD.

### 6) forgot_password
Send an email to the user who requests to reset the password. 
When the user clicks the link in the email, web client will send this user a new password with email.
The UI will be handled by web client.
Android client should go to log in activity.

#### Request
| Name   | Description                 |
| ------ | --------------------------- |
| Method | POST                        |
| URL    | `api/resetPassword`         |
| DATA   | `email`=abc@def.com         |

#### Response
```javascript
{
    "status": "success",
    "expiration": "2014-03-06T03:01:37+00:00",
    "data": {
        "message": "An email containing the steps to reset password has been sent to your email account."    
    }
}
```

#### Associated Errors
* 1 - Please enter your email address
* 2 - Please try this operation later (too many requests in one session period)
* 3 - Invalid email address
* 4 - Email not registered
* 5 - Email did not send due to server error
* 6 - Email did not send due to server runtime error

### 7) Get Profile of a User

Get the profile of the user.

__Last Update__: March 21, 2014

#### Request
| Name   | Description                       |
| ------ | --------------------------------- |
| Method | POST                              |
| URL    | `api/user/info/@target_user_id`   |
| DATA   | `user_id`=123&`ugl_token`=mytoken |

* `user_id` and `ugl_token` are the credentials to verify the requester.
* `@target_user_id` in the URL is the ID of the user to fetch profile.

#### Response

Upon success, server will return something like
```javascript
{
    "status": "success",
    "expiration": "2014-03-21T04:07:32+00:00",
    "data": {
        "id": "5",
        "email": "xb@purdue.edu",
        "nickname": null,
        "first_name": "X",
        "last_name": "B",
        "avatar_url": "",
        "phone": null,
        "description": null,
        "created_at": "2014-03-17 11:13:55",
        "_preferences": {
            "autoAcceptInvitation": false,
            "showMyProfile": true,
            "showMyPublicGroups": true
        }
    }
}
```

* Field `_preferences` will show only if the requester is the target user.
* When fetching others' profile, if the target user does not allow others to view his / her profile, an error will occur.

#### Associated Errors
* 1 - You should log in to perform the request (At least one of POST fields `user_id` and `ugl_token` is missing)
* 2 - Unauthorized request (`user_id` and `ugl_token` do not match the user)
* 3 - Invalid user id (`@target_user_id` is not numerical)
* 4 - User not found (`@target_user_id` does not exist)
* 5 - The profile is set private (target user does not allow you to view his / her profile)

### 8) Set profile for a user

Update the profile and preferences for the requester.

To change the password for a user, use the API entry named `change_password`.

__Last Update__: March 21, 2014

#### Request

| Name   | Description                       |
| ------ | --------------------------------- |
| Method | POST                              |
| URL    | `/api/user/edit`                  |
| DATA   | `user_id`=123&`ugl_token`=mytoken&... |

* `...` stands for `email`=new_email&`first_name`=aaaa&`last_name`=bbbb&`nickname`=nick&`avatar_url`=blah&`phone`=123&`description`=oops&`autoAcceptInvitation`=true|false&`showMyProfile`=true|false&`showMyPublicGroups`=true|false
* `email`, `first_name` and `last_name` are required
	 * `email`, if changed, must be an email that is not registered yet
	 * `first_name` and `last_name` are non-empty string
* `nickname`, `avatar_url`, `phone`, if un-POSTed, will be changed to empty string
	 * note: empty string for `avatar_url` indicates to use the default one
* `description` will be HTML-encoded (see the description field for Group identity)
	 * if un-POSTed, `description` will remain unchanged to save data transmission for the client
* For every preference field, simply POST in the format `key=value`.

#### Response

Upon success, the new user identity will be returned (see section [Get profile of a user](#7-get-profile-of-a-user) with the requester being the target user).

#### Associated Errors
* 1 - You should log in to perform the request (At least one of POST fields `user_id` and `ugl_token` is missing)
* 2 - Unauthorized request (`user_id` and `ugl_token` do not match the user)
* 3 - Invalid email address
* 4 - Email already registered
* 5 - First name or last name should be non-empty words

### 9) Change password

Change the password of a user, provided that the user knows the original password.

### 10) Upload Avatar

***

## 2. Group API

The data fields for group model are defined as below:

* **id**: the unique ID for the group.
* **alias**: A user-defined identifier of the group. It is a __non_empty__ string of 
	length no greater than 32, consisting of only *letters*, *digits*, '*-*', and '*_*' 
	(any match of `[^\-_a-zA-Z0-9]` should make the string invalid as group name).
* **description**: The description (brief introduction) of the group. All HTML escape 
	characters should be escaped by the client (when displaying the field, strings inside 
	"<" and ">" must no execute in face of XSS attacks, etc.). and the after-filtering string 
	has a maximum length of *150* chars. (Note: as a result, for example, "&", equivalent to "&amp;", has a length of 5.)
* **status**: 
	 * `0`: The group is closed (deleted)
	 * `1`: The group is inactive (users can access info but cannot change it; admin can re-active it)
	 * `2`: The group is private (invisible to outsiders)
	 * `3`: The group is public (data accessible to everyone; specific permissions are decided by the role of the requester)
* **creator_user_id**: The user id of the creator
* **num_of_users**: The number of users in the group
* **avatar_url**: The URL of the group avatar image. Empty means using default one.
* **users**: a list of users in the group, categorized by their roles
* **_preferences**: (Private field) the group preference settings
	 * `autoApproveApplication`: if set to `1`, the group will add applicants directly to `member` role. Otherwise he will be added to `pending` role.

### 1) List Groups Of
List the groups a user has joined.

* When the requester is not the "target user" (the one whose groups to be listed), then upon this user's preference, either only his / her public groups are listed, or no groups are shown (Refer to `showMyPublicGroups` preference).
* When the requester is the target user, all his / her groups will be listed.

#### Request
| Name   | Description                                             |
| ------ | ------------------------------------------------------- |
| Method | POST                                                    |
| URL    | `/api/user/listGroup/@target_user_id`                   |
| DATA   | `user_id`=123&`ugl_token`=mytoken                       |

In the URL, `@target_user_id` is the id of the user whose groups are to be listed. For example, to get the groups of user whose id is `444`, the URL should be `api/listGroupsOf/444`.

`user_id` and `ugl_token` are the log in credentials of the requester. No guest can perform the operation.

#### Response
The API response to `/api/listGroupsOf/2` looks like (more fields may be added in the future)

```javascript
{
    "status": "success",
    "expiration": "2014-03-17T07:25:07+00:00",
    "data": {
        "count": 3,
        "groups": [
            {
                "id": "1",
                "status": "1",
                "alias": "admin",
                "description": "",
                "avatar_url": null,
                "tags": null,
                "creator_user_id": "3",
                "num_of_users": "1",
                "users": {
                    "admin": [
                        "2"
                    ]
                },
                "created_at": "0000-00-00 00:00:00"
            },
            {
                "id": "2",
                "status": "2",
                "alias": "ugl-dev",
                "description": "UGL Developers",
                "avatar_url": null,
                "tags": null,
                "creator_user_id": "2",
                "num_of_users": "1",
                "users": {
                    "admin": [
                        "2"
                    ],
                    "member": [
                        "3"
                    ]
                },
                "created_at": "0000-00-00 00:00:00"
            },
            {
                "id": "5",
                "status": "3",
                "alias": "pucs",
                "description": "&lt;script&gt;\r\n\talert(&quot;hello!&quot;)\r\n&lt;\/script&gt;\r\n&lt;b&gt;Abercrombie &amp; Fitch&lt;\/b&gt;",
                "avatar_url": null,
                "tags": "purdue cs test",
                "creator_user_id": "2",
                "num_of_users": "1",
                "users": {
                    "admin": [
                        "2"
                    ]
                },
                "created_at": "2014-03-17 06:07:30"
            }
        ]
    }
}
```

Notes:
* `data.count` is the number of groups in the list, and `data.groups` is the list of groups.
* Notice how the `description` is HTML encoded for security's sake. A client may need to decode it to display the text.
* `users` is a sub-array whose structure is `role` => `members`.
* Group preference fields will not list in the array. To get the group preference, the request must be an admin requesting the group info (see getInfo API).

#### Associated Errors
* 1 - You should log in to perform the request (At least one of POST fields `user_id` and `ugl_token` is missing)
* 2 - Unauthorized request (`user_id` and `ugl_token` do not match the user)
* 3 - User id should be a number (`@target_user_id` is not numerical)
* 4 - The user does not exist (user whose id is `@target_user_id` does not exist)

### 2) Create a Group
Create a new group.

#### Request
| Name   | Description                                             |
| ------ | ------------------------------------------------------- |
| Method | POST                                                    |
| URL    | `/api/group/create`                                     |
| DATA   | `user_id`=123&`ugl_token`=mytoken&`alias`=group_name&`description`=group_description&`tags`=group_tags&`status`=1  |

URLEncode the fields when necessary.

#### Response

Upon success, server will return a success message, and the filtered group data:

* `description` will be HTML encoded and cut to the max length `150`.
* `tags` will be removed all neither-alphanumerical-nor-space characters and duplicate words, and then the tags get sorted`

```javascript
{
    "status": "success",
    "expiration": "2014-03-17T07:31:41+00:00",
    "data": {
        "message": "Successfully created a new group",
        "group_data": {
            "id": "12",
            "status": "2",
            "alias": "test-animation-effect",
            "description": "hello!\r\nLet's see the bounce!!!",
            "avatar_url": null,
            "tags": "delete notag oops purdue test",
            "creator_user_id": "2",
            "num_of_users": "1",
            "users": {
                "admin": [
                    "2"
                ]
            },
            "created_at": "2014-03-17 07:31:41"
        }
    }
}
```

#### Associated Errors

* 1 - You should log in to perform the request (At least one of POST fields `user_id` and `ugl_token` is missing)
* 2 - Unauthorized request (`user_id` and `ugl_token` do not match the user)
* 3 - Group name is not of the specified format. Plese check
* 4 - Group name "blah" is already taken
* 5 - Please choose a valid status option from the list


### 3) Delete or Leave a Group or Kick Members

This API combines two operations of leaving and deleting a group into one, given the different combinations of parameters.

#### Request
| Name   | Description                               |
| ------ | ----------------------------------------- |
| Method | POST                                      |
| URL    | `/api/group/leave`                        |

* POST data `user_id`=123&`ugl_token`=mytoken&`group_id`=target_group_id
* When the creator leaves the group, the group will be removed (deleted)
	 * He / she can choose to notify all group members of the shutdown by POSTing a parameter "`notify`=true"
* When an admin (or whoever has "manage" permission) wants to kick a group user other than himself, also POST a parameter "`target_user_id`=user_to_kick"
* No one can kick the creator unless he / she is the creator, but the creator doing so means deleting the group
* Uncaught invalid requests will remove the requester from the group if he is in it
* Kicking unregistered user will be ignored by server, but the client will get a success message
* To kick more than one users from the group, POST `target_user_id` parameter like "`target_user_id`=1,2,44,7" (this will kick users whose IDs are in the list of the group)
	 * if the creator's ID is in the list, he / she will NOT get kicked
	 * if an user ID is unregistered or is not in the group, kicking this user makes NO changes to the group
	 * invalid IDs in the list (e.g., `aa` as in `target_user_id=aa,123`) will be ignored, the rest of the operation will be performed
* An admin kicking himself will be considered as a "kick" operation instead of a "leave" operation
	 
#### Response
Upon success, the requester will receive one of the following depending on the particular request

* `The group is now closed` (The creator deleted the group)
* `You have successfully kicked the user` (an admin kicked some user)
* `You have successfully left the group` (a user left the group)

#### Associated Errors

* 1 - You should log in to perform the request (At least one of POST fields `user_id` and `ugl_token` is missing)
* 2 - Unauthorized request (`user_id` and `ugl_token` do not match the user)
* 3 - Group id not specified (`group_id` is not POSTed)

CHANGELOG.md
# Sprint 2

### Mar 15, 2014
* News model to record user events, rev.1

### Mar 14, 2014
* Group model, controller, API, and UI rev.1
* App-side OAuth callback API

### Mar 13, 2014
* Fully functioning user password reset route
* A beautiful 404 page
* Added logging mechanism for Controller

### Mar 12, 2014
* Brand-new AJAX-based, responsive, and animate user control panel

### Mar 11, 2014
* Improved logOut API
* Improved User model for performance and security
* Added Mail model
* Added a generic Preference model
* Bug fixes

### Mar 10, 2014
* New theme

### Mar 07, 2014
* Clean ups


app/models/Group.php
<?php
/**
 * The group model used in the system
 *
 * @author	Xiangyu Bu
 * @version	0.2
 */

namespace models;

class Group extends \Model {
	
	const MAX_ALIAS_LENGTH = 32;
	const MAX_DESC_LENGTH = 150;
	const GROUP_RECORD_TTL = 1800; // in sec
	
	const STATUS_CLOSED = 0; // appears non-existent to users
	const STATUS_INACTIVE = 1; // users can still access the group data, but cannot change it 
	const STATUS_PRIVATE = 2; // private groups are invisible to outsiders and are invitation-only
	const STATUS_PUBLIC = 3; // everyone can see the group
	
	// group-wide (minimum) permissions
	static $DEFAULT_GROUP_PERMISSION = array(
		"role_name" => "guest",
		"view_profile" => false,	// view group profile
		"apply" => false, 		// apply to join
		"view_board" => false,	// view the boards
		"new_board" => false,	// new board
		"edit_board" => false,	// edit a board
		"del_board" => false,	// delete a board
		"post" => false, 	// post subjects
		"comment" => false, // comment on posts
		"delete" => false,	// delete posts
		"edit" => false, 	// edit subjects
		"manage" => false 	// manage members and change profile
	);
	
	// default group preferences
	static $DEFAULT_GROUP_PREFS = array(
		"autoApproveApplication" => 0
	);
	
	/**
	 * isValidGroupName
	 * Check if the string contains chars other than alphanumerical, -, _
	 * and non-empty, and length not greater than 65.
	 */
	static function isValidAlias($str){
		if (empty($str)) return false;
		if (preg_match('/[^\-_A-Za-z0-9]/', $str)) return false;
		if (strlen($str) > static::MAX_ALIAS_LENGTH) return false;
		return true;
	}
	
	static function isValidStatus($v){
		return is_numeric($v) and $v > static::STATUS_CLOSED and $v <= static::STATUS_PUBLIC;
	}
	
	static function filterDescription($str){
		$str = htmlspecialchars($str);
		$str = substr($str, 0, static::MAX_DESC_LENGTH);
		return $str;
	}
	
	static function filterTags($str){
		// remove non alphanumerical chars or delimiters
		$str = preg_replace("/[^A-Za-z0-9 ]/", "", strtolower($str));
		// remove repeated words
		$str = preg_replace("/\b(\w+)\s+\\1\b/i", "$1", $str);
		$str_a = explode(" ", $str);
		sort($str_a); // sort the words
		$str = implode(" ", $str_a);
		return $str;
	}
	
	function findById($id){
		if ($this->cache->exists("group_id_" . $id))
			return $this->cache->get("group_id_" . $id);
		
		$result = $this->queryDb("SELECT * FROM groups WHERE id=? AND status > " . static::STATUS_CLOSED, $id);
		if (count($result) == 1){
			$result = $result[0];
			$result["users"] = json_decode($result["users"], true);
			if ($result["_preferences"]) $result["_preferences"] = json_decode($result["_preferences"], true);
			else $result["_preferences"] = self::$DEFAULT_GROUP_PREFS;
			
			$this->cache->set("group_id_" . $id, $result, static::GROUP_RECORD_TTL);
		} else $result = null;
		
		return $result;
	}
	
	function findByAlias($alias){
		$result = $this->queryDb("SELECT id FROM groups WHERE alias=? AND status > " . static::STATUS_CLOSED, $alias);
		if (count($result) == 1)
			return $this->findById($result[0]["id"]);
		return null;
	}
	
	// public only
	function findByKeyword($keyword){
		$ids = $this->queryDb("SELECT id FROM groups WHERE status=" . static::STATUS_PUBLIC . " AND CONCAT_WS(' ', alias, description, tags) LIKE '?';", "%" . $keyword . "%", 7200);
		if (count($ids) > 0){
			$result = array();
			foreach ($ids as $k => $id) $result[] = $this->findById($id);
			return $result;
		}
		return null;
	}
	
	// corresponding to static::STATUS_INACTIVE
	function listGroupsOfUserId($user_id, $target_status = 1){
		$ids = $this->queryDb("SELECT id FROM groups WHERE status >= " . $target_status . " AND users LIKE '%\"" . $user_id . "\"%' ORDER BY status DESC");
		$result = array();
		foreach ($ids as $i => $d)
			$result[] = $this->findById($d["id"]);
		
		if (count($result) > 0)
			return array("count" => count($result), "groups" =>$result);
		return array("count" => 0);
	}
	
	function getDefaultRoleName(){
		return "member";
	}
	
	function getPendingRoleName(){
		return "pending";
	}
	
	function getInviteeRoleName(){
		return "invitee";
	}
	
	function getPermissions($user_id, $group_id, $group_data = null){
		
		$permissions = self::$DEFAULT_GROUP_PERMISSION;
		if (!$group_data) $group_data = $this->findById($group_id);
		
		if (array_key_exists("admin", $group_data["users"]) and $user_id > 0 and in_array($user_id, $group_data["users"]["admin"])){
			$permissions["role_name"] = "admin";
			$permissions["view_profile"] = true;
			$permissions["view_board"] = true;
			if ($group_data["status"] > static::STATUS_INACTIVE){
				$permissions["new_board"] = true;
				$permissions["edit_board"] = true;
				$permissions["del_board"] = true;
				$permissions["post"] = true; 
				$permissions["comment"] = true; 
				$permissions["delete"] = true;
			}
			$permissions["edit"] = true;
			$permissions["manage"] = true;
		} else if (array_key_exists("member", $group_data["users"]) and $user_id > 0 and in_array($user_id, $group_data["users"]["member"])){
			$permissions["role_name"] = "member";
			$permissions["view_profile"] = true;
			$permissions["view_board"] = true;
			if ($group_data["status"] > static::STATUS_INACTIVE){
				$permissions["new_board"] = true;
				$permissions["edit_board"] = true;
				$permissions["post"] = true;
				$permissions["comment"] = true;
			}
		} else {
			// invitees and applicants are treated as guests
			
			if ($group_data["status"] > static::STATUS_PRIVATE)
				$permissions["view_profile"] = true;
			
			if ($user_id > 0 and array_key_exists("pending", $group_data["users"]) and in_array($user_id, $group_data["users"]["pending"])){
				$permissions["role_name"] = "pending";
			} else if ($user_id > 0 and $group_data["status"] > static::STATUS_PRIVATE)
				// of course a guest cannot apply
				$permissions["apply"] = true;
		}
		
		return $permissions;
	}
	
	function create($user_id, $alias, $desc, $tags, $status){
		$this->queryDb(
			"INSERT INTO groups (creator_user_id, status, alias, description, tags, users, created_at) " .
			"VALUES (:user_id, :status, :alias, :desc, :tags, :users, NOW()); ",
			array(
				':user_id' => $user_id,
				':status' => $status,
				':alias' => $alias,
				':desc' => $desc,
				':tags' => $tags,
				':users' => json_encode(array("admin" => array($user_id)))
			)
		);
		
		return $this->findByAlias($alias);
	}
	
	function update(&$group_data, $alias, $desc, $tags, $status){
		$changed = false;
		
		if ($alias != $group_data["alias"]){
			$changed = true;
			$group_data["alias"] = $alias;
		}
		
		if ($desc != $group_data["description"]){
			$changed = true;
			$group_data["description"] = $desc;
		}
		
		if ($tags != $group_data["tags"]){
			$changed = true;
			$group_data["tags"] = $tags;
		}
		
		if ($status != $group_data["status"]){
			$changed = true;
			$group_data["status"] = $status;
		}
		
		if ($changed){
			$this->save($group_data);
			if ($this->cache->exists("group_id_" . $group_data["id"]))
				$this->cache->set("group_id_" . $group_data["id"], $group_data);
			//$this->cache->clear("group_id_" . $group_data["id"]);
		}
	}
	
	function save($group_data){
		$this->queryDb("UPDATE groups " .
			"SET status=:status, alias=:alias, description=:description, avatar_url=:avatar_url, tags=:tags, creator_user_id=:creator_user_id, users=:users, _preferences=:prefs ".
			"WHERE id=:id;",
			array(
				":id" => $group_data["id"],
				":status" => $group_data["status"],
				":alias" => $group_data["alias"],
				":description" => $group_data["description"],
				":avatar_url" => $group_data["avatar_url"],
				":tags" => $group_data["tags"],
				":creator_user_id" => $group_data["creator_user_id"],
				":users" => json_encode($group_data["users"]),
				":prefs" => json_encode($group_data["_preferences"]),
			)
		);
	}
	
	function deleteById($gid){
		//TODO: delete all records related to the group before deleting it

		$this->queryDb("UPDATE groups SET status=:status WHERE id=:id", array(":status" => static::STATUS_CLOSED, ":id" => $gid));
		$this->cache->clear("group_id_" . $gid);
	}
	
	function setCreatorUserId($uid, &$group_data){
		$group_data["creator_user_id"] = $uid;
		
		if ($this->cache->exists("group_id_" . $group_data["id"]))
			$this->cache->set("group_id_" . $group_data["id"], $group_data);
	}
	
	function addUser($uid, $role, &$group_data){
		$group_data["users"][$role][] = $uid;
		
		if ($this->cache->exists("group_id_" . $group_data["id"]))
			$this->cache->set("group_id_" . $group_data["id"], $group_data);
	}
	
	function kickUser($uid, &$group_data){
		$flag = false;
		
		foreach ($group_data["users"] as $role => $users){
			$i = array_search("" . $uid . "", $users);
			if ($i) {
				unset($group_data["users"][$role][$i]);
				$flag = true;
			}
		}
		
		if ($flag and $this->cache->exists("group_id_" . $group_data["id"]))
			$this->cache->set("group_id_" . $group_data["id"], $group_data);
		
		return $flag;
	}
	
	function changeUserRole($uid, $role, &$group_data){
		$this->kickUser($uid, $group_data);
		$group_data["users"][$role][] = "" . $uid . "";
		
		if ($flag and $this->cache->exists("group_id_" . $group_data["id"]))
			$this->cache->set("group_id_" . $group_data["id"], $group_data);
	}
}